apm-agent-core/src/main/java/co/elastic/apm/agent/universalprofiling/UniversalProfilingIntegration.java

```diff
@@ -70,21 +95,49 @@ public void start(ElasticApmTracer tracer) {
         try {
             log.debug("Starting universal profiling correlation");
 
+            socketPath = openProfilerSocket(config.getSocketDir());
+
             CoreConfiguration coreConfig = tracer.getConfig(CoreConfiguration.class);
             ByteBuffer processCorrelationStorage = ProfilerSharedMemoryWriter.generateProcessCorrelationStorage(
-                coreConfig.getServiceName(), coreConfig.getEnvironment(), "");
+                coreConfig.getServiceName(), coreConfig.getEnvironment(), socketPath);
             UniversalProfilingCorrelation.setProcessStorage(processCorrelationStorage);
 
+            executor = ExecutorUtils.createSingleThreadSchedulingDaemonPool("profiling-integration");
+            executor.scheduleWithFixedDelay(new Runnable() {
+                @Override
+                public void run() {
+                    periodicTimer();
+                }
+            }, POLL_FREQUENCY_MS, POLL_FREQUENCY_MS, TimeUnit.MILLISECONDS);
+
             isActive = true;
             tracer.registerSpanListener(activationListener);
         } catch (Exception e) {
             log.error("Failed to start universal profiling integration", e);
+            if (socketPath != null) {
+                try {
+                    UniversalProfilingCorrelation.stopProfilerReturnChannel();
+                    socketPath = null;
+                } catch (Exception e2) {
+                    log.error("Failed to clean up universal profiling integration socket", e2);
+                }
+            }
         }
     }
 
+    private void periodicTimer() {
+        consumeProfilerMessages();
+    }
+
     public void stop() {
         try {
+            if (executor != null) {
+                executor.shutdown();
+                executor.awaitTermination(10, TimeUnit.SECONDS);
+                executor = null;
+            }
             if (isActive) {
+                UniversalProfilingCorrelation.stopProfilerReturnChannel();
```

Comment:
do you need to set isActive false at scope exit?

Reply:
IIRC, the object won't be reused after a call to `stop()`. Nonetheless, it looks more clean to me to reset it.